File: 3/SharedPrefix.java
class SharedPrefix {
  public String biggestSharedPrefix(String[] gifts) {
    int numOfGifts = gifts.length;

    if (numOfGifts == 0)
      return "";

    int shortestGift = 99999999;

    for (int i = 0; i < numOfGifts; ++i)
      shortestGift = Math.min(shortestGift, gifts[i].length());

    for (int j = 0; j < shortestGift; ++j)
    {
      for (int i = 1; i < numOfGifts; ++i)
      {
        if (!(gifts[0].charAt(j) == gifts[i].charAt(j)))
          return gifts[0].substring(0, j);
      }
    }

    return gifts[0].substring(0, shortestGift);
  }
}

File: 3/SharedPrefix.kt
class SharedPrefix {
  fun biggestSharedPrefix(gifts: Array<String>): String {
    var biggestSharedPrefix = ""

    var currentCharacter = ""
    var currentIndex = 0

    while(true) {
      for(element in gifts) {
        var giftName = element

        if(giftName.length < currentIndex) {
          return biggestSharedPrefix
        }

        var giftNameCharacter = giftName.toCharArray()[currentIndex].toString()
        if(currentCharacter.isBlank()) {
          currentCharacter = giftNameCharacter
        } else if(currentCharacter != giftNameCharacter) {
          return biggestSharedPrefix
        }
      }

      biggestSharedPrefix += currentCharacter
      currentCharacter = ""
      currentIndex++
    }

    return biggestSharedPrefix
  }
}
